### PR TITLE
UnitTesting and Razor External Access wrappers for ISB services

### PR DESCRIPTION
**Pull request by tmat**
Corresponding changes in UnitTesting: https://devdiv.visualstudio.com/DevDiv/_git/VSUnitTesting/pullrequest/282266

Fixes https://github.com/dotnet/roslyn/issues/44327

### CODE REVIEW

**Comment by tmat** [comment]
@shyamnamboodiripad PTAL

**Comment by tmat** [comment]
@dotnet/roslyn-ide

File: src/Tools/ExternalAccess/Razor/Remote/RazorRemoteServiceConnectionWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Remote;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal readonly struct RazorRemoteServiceConnectionWrapper<TService> : IDisposable
```
Comment (sharwell) [inline src/Tools/ExternalAccess/Razor/Remote/RazorRemoteServiceConnectionWrapper.cs:12]: ❔ Why are we making a wrapper if it's not part of the RIVT API?

File: src/Workspaces/Core/Portable/Remote/IRemoteServiceCallbackDispatcherProvider.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal interface IRemoteServiceCallbackDispatcherProvider
+    {
+        IRemoteServiceCallbackDispatcher GetDispatcher(Type serviceType);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Remote/IRemoteServiceCallbackDispatcherProvider.cs:11]: 😕 This seems like something the partner would manage as a completely independent brokered service. The only part they need from us is a brokered service that gives access to a specific solution.

File: src/Tools/ExternalAccess/Razor/Remote/RazorRemoteServiceConnectionWrapper.cs (hunk 1 above)
Comment (tmat) [reply src/Tools/ExternalAccess/Razor/Remote/RazorRemoteServiceConnectionWrapper.cs:12]: This is not RIVT but a separate assembly.

File: src/Workspaces/Core/Portable/Remote/IRemoteServiceCallbackDispatcherProvider.cs (hunk 2 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Remote/IRemoteServiceCallbackDispatcherProvider.cs:11]: We could remove these but then everyone would need to reimplement pretty much the same logic we have in Roslyn. I opted to make that infrastructure available, so it's easy to use ISB for partners.

Comment (tmat) [reply src/Workspaces/Core/Portable/Remote/IRemoteServiceCallbackDispatcherProvider.cs:11]: We also don't want to expose the different descriptors for 32/64/64S.

File: src/Workspaces/Remote/Core/ServiceDescriptors.cs [hunk 3]
```diff
@@ -99,7 +126,29 @@ public static ServiceDescriptor GetServiceDescriptor(Type serviceType, bool isRe
...
         }
 
-        internal static string GetFeatureName(Type serviceInterface)
-            => RemoteWorkspacesResources.GetResourceString("FeatureName_" + GetServiceName(serviceInterface));
+        internal static string GetFeatureDisplayName(string qualifiedServiceName)
+        {
+            var prefixLength = qualifiedServiceName.LastIndexOf('.') + 1;
+            Contract.ThrowIfFalse(prefixLength > 0);
+
+            var suffixLength = qualifiedServiceName.EndsWith(RemoteServiceName.Suffix64, StringComparison.Ordinal) ? RemoteServiceName.Suffix64.Length : 0;
```
Comment (sharwell) [inline src/Workspaces/Remote/Core/ServiceDescriptors.cs:134]: 📝 This needs to be updated to also support the server GC suffix.

File: src/Workspaces/CoreTest/Remote/ServiceDescriptorTests.cs [hunk 4]
```diff
@@ -163,10 +164,12 @@ public void TypesUsedInRemoteApisMustBeMessagePackSerializable()
...
         [Theory]
-        [MemberData(nameof(ServiceTypes))]
-        public void GetFeatureName(Type serviceType)
+        [MemberData(nameof(AllServiceDescriptors))]
+        internal void GetFeatureDisplayName(Type serviceInterface, ServiceDescriptor descriptor32, ServiceDescriptor descriptor64)
         {
-            Assert.NotEmpty(ServiceDescriptors.GetFeatureName(serviceType));
+            Assert.NotNull(serviceInterface);
+            Assert.NotEmpty(descriptor32.GetFeatureDisplayName());
+            Assert.NotEmpty(descriptor64.GetFeatureDisplayName());
```
Comment (sharwell) [inline src/Workspaces/CoreTest/Remote/ServiceDescriptorTests.cs:172]: 📝 Need to test the server GC descriptor as well

**Review by sharwell — APPROVED** [review]
Overall looks fine. Eventually I'd like to see the IVT dependencies here reduced, first by eliminating the use of Roslyn's callback dispatchers (these can be implemented by each partner separately, and if it's too complex then ServiceHub itself needs to provide this functionality), and then by moving the services to independent brokered services that depend only on Roslyn's brokered service that provides access to the workspace and/or solution.

We need to make sure this works with Server GC scenarios as well.

File: src/Workspaces/Remote/Core/ServiceDescriptors.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/Remote/Core/ServiceDescriptors.cs]: ➡️ Now updated and tested

File: src/Workspaces/CoreTest/Remote/ServiceDescriptorTests.cs (hunk 4 above)
Comment (sharwell) [reply src/Workspaces/CoreTest/Remote/ServiceDescriptorTests.cs]: ➡️ Done

File: src/Tools/ExternalAccess/Razor/Remote/RazorRemoteHostClient.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+using System.Threading.Tasks;
+using MessagePack;
+using MessagePack.Formatters;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Remote;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor
+{
+    internal readonly struct RazorRemoteHostClient
```
Comment (genlu) [inline src/Tools/ExternalAccess/Razor/Remote/RazorRemoteHostClient.cs:19]: Changing this from class to struct is a breaking change

Comment (genlu) [reply src/Tools/ExternalAccess/Razor/Remote/RazorRemoteHostClient.cs:19]: FYI @NTaylorMullen

**Pull request merged**